### PR TITLE
Add var_conv_2d op

### PR DESCRIPTION
**Pull request by Aurelius84**
- **var_con_2d**: support conv2d of input sequence with variable length
Example:
```python
import numpy as np
from paddle.fluid import layers
x_lod_tensor = layers.data(name='x', shape=[1], lod_level=1)
row_lod_tensor = layers.data(name='row', shape=[6], lod_level=1)
col_lod_tensor = layers.data(name='col', shape=[6], lod_level=1)
out = layers.var_conv_2d(input=x_lod_tensor, 
                         row=row_lod_tensor,
                         col=col_lod_tensor,
                         input_channel=3,
                         output_channel=5,
                         filter_size=[3, 3],
                         stride=1)
```
var_conv2d

### CODE REVIEW

File: paddle/fluid/operators/CMakeLists.txt [hunk 1]
```diff
@@ -48,8 +48,13 @@ if (WITH_DISTRIBUTE)
     SET(OP_PREFETCH_DEPS ${OP_PREFETCH_DEPS} parameter_prefetch)
 endif()
 
+SET(OP_ONLY_MKL "")
+if (NOT WITH_MKL)
+    SET(OP_ONLY_MKL ${OP_ONLY_MKL} var_conv_2d_op)
+endif()
```
Comment (luotao1) [inline paddle/fluid/operators/CMakeLists.txt:54]: line51-54 is not elegant.
if (NOT WITH_MKL), `var_conv_2d_op` should not be compiled and added into `GLOB_OP_LIB `

Maybe you should write the following instead.
```
if (WITH_MKL)
   op_library(var_conv_2d_op) 
endif()
```

Comment (Aurelius84) [reply paddle/fluid/operators/CMakeLists.txt:54]: I tried your code, but it seems not work. Also, I tried this, still don't work:
```
if (WITH_MKL)
    op_library(var_conv_2d_op)
    file(APPEND ${pybind_file} "ONLY_AVLID_WITH_MKL(var_conv_2d);\n")
endif()
```

File: paddle/fluid/operators/var_conv_2d_op.cc [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    auto* col = ctx.Input<LoDTensor>("Col");
+    auto* out = ctx.Input<LoDTensor>("Out");
+
+    int output_channel = ctx.Attr<int>("OutputChannel");
+    int input_channel = ctx.Attr<int>("InputChannel");
+    int kernel_h = ctx.Attr<int>("KernelH");
+    int kernel_w = ctx.Attr<int>("KernelW");
+
+    auto* d_out = ctx.Input<LoDTensor>(framework::GradVarName("Out"));
+    // auto* d_col = ctx.Input<LoDTensor>(framework::GradVarName("Col"));
```
Comment (lanxianghit) [inline paddle/fluid/operators/var_conv_2d_op.cc]: please remove this

File: paddle/fluid/operators/var_conv_2d_op.cc [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+    auto* dx = ctx.Output<LoDTensor>(framework::GradVarName("X"));
+    auto* d_w = ctx.Output<Tensor>(framework::GradVarName("W"));
+
+    Tensor col_grad;
+    col_grad.Resize(col->dims());
+    auto* col_diff = col_grad.mutable_data<T>(ctx.GetPlace());
+    auto* dx_data = dx->mutable_data<T>(ctx.GetPlace());
+    auto* w_diff = d_w->mutable_data<T>(ctx.GetPlace());
+    // auto* col_diff = const_cast<T*>(d_col->data<T>());
```
Comment (lanxianghit) [inline paddle/fluid/operators/var_conv_2d_op.cc]: remove

**Review by lanxianghit — CHANGES_REQUESTED** [review]

**Review by lanxianghit — DISMISSED** [review]

**Review by lanxianghit — DISMISSED** [review]

File: paddle/fluid/operators/var_conv_2d_op.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline paddle/fluid/operators/var_conv_2d_op.h:1]: 2018 -> 2019

**Review by heavengate — DISMISSED** [review]

**Review by lanxianghit — APPROVED** [review]

**Review by heavengate — APPROVED** [review]

**Review by luotao1 — APPROVED** [review]
LGTM for `op_use_default_grad_op_maker.spec`

**Pull request merged**